File: Python/BOJ/24480 알고리즘 수업 - 깊이 우선 탐색 2.py
import sys
sys.setrecursionlimit(200000)

N, M, R = map(int, sys.stdin.readline().split())
graph = [[] for _ in range(N+1)]
for _ in range(M):
    u, v = map(int, sys.stdin.readline().split())
    graph[u].append(v)
    graph[v].append(u)
visited = [0]*(N+1)
cnt = 1

def dfs(n):
    global cnt
    visited[n] = cnt
    graph[n].sort(reverse=True)
    for i in graph[n]:
        if visited[i] == 0:
            cnt += 1
            dfs(i)
dfs(R)
for i in range(1,N+1):
    print(visited[i])

File: Java/BOJ/24480 알고리즘 수업 - 깊이 우선 탐색 2.java
import java.util.*;
import java.io.*;

public class Main {
	static int N, M, R;
	static ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
	static int[] visited;
	static int cnt = 1;
	
	public static void dfs(int n) {
		visited[n] = cnt;
		Collections.sort(graph.get(n), Comparator.reverseOrder());
		for(int i = 0; i < graph.get(n).size(); i++) {
			int x = graph.get(n).get(i);
			if (visited[x] == 0) {
				cnt ++;
				dfs(x);
			}
		}
	}
	
    public static void main(String[] args) throws IOException {
    	BufferedReader br =  new BufferedReader(new InputStreamReader(System.in));
    	StringTokenizer st = new StringTokenizer(br.readLine());
    	N = Integer.parseInt(st.nextToken());
    	M = Integer.parseInt(st.nextToken());
    	R = Integer.parseInt(st.nextToken());
    	visited = new int[N+1];
    	for(int i = 0; i < N+1; i++) {
        	graph.add(new ArrayList<>());
        }
    	for (int i = 0 ; i < M; i++) {
    		st = new StringTokenizer(br.readLine());
    		int u = Integer.parseInt(st.nextToken());
    		int v = Integer.parseInt(st.nextToken());
    		graph.get(u).add(v);
    		graph.get(v).add(u);
    	}
    	
    	dfs(R);
    	StringBuilder sb = new StringBuilder();
    	for (int i = 1; i < N+1; i++) {
        	sb.append(visited[i]).append("\n");
        }
        System.out.println(sb);
    }
}